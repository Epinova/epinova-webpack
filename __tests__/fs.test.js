/**
 * @jest-environment node
 */

const webpack = require("webpack");
const { createFsFromVolume, Volume } = require("memfs");
const joinPath = require("memory-fs/lib/join");
const path = require("path");
const serializer = require("jest-serializer-path");

const config = require("../config");

expect.addSnapshotSerializer(serializer);

function ensureWebpackMemoryFs(fs) {
    // Return it back, when it has Webpack 'join' method
    if (fs.join) {
        return fs;
    }

    // Create FS proxy, adding `join` method to memfs, but not modifying original object
    const nextFs = Object.create(fs);
    nextFs.join = joinPath;

    return nextFs;
}

// Usage with Webpack:

function buildWebpackCompiler(fs, webpackConfig) {
    const webpackFs = ensureWebpackMemoryFs(fs);

    const compiler = webpack(webpackConfig);

    compiler.outputFileSystem = webpackFs;
    compiler.resolvers.context.fileSystem = webpackFs;

    return compiler;
}
test("fs", async () => {
    Date.now = jest.fn(() => 1482363367071);

    const vol = new Volume();
    const fs = new createFsFromVolume(vol);
    fs.mkdirSync("/dist");

    const promise = new Promise((resolve, reject) => {
        const c = config({ path: "dist", outputPath: "/dist" }, config => {
            config.entry = path.join(__dirname, "../example/app.js");

            config.optimization = {};

            config.plugins = [];

            return config;
        })("development", { mode: "development" });

        const compiler = buildWebpackCompiler(fs, c);
        return compiler.run((err, stats) => {
            if (err) {
                return reject(err);
            }

            resolve(stats);
        });
    });

    return promise.then(stats => {
        expect(
            stats.toJson({
                assets: true,
                hash: true,
                modules: false
            })
        ).toMatchSnapshot();
        expect(vol.toJSON()).toMatchSnapshot();
    });
});
