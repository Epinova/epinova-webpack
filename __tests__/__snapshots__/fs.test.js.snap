// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fs 1`] = `
Object {
  "assets": Array [
    Object {
      "chunkNames": Array [
        "main",
      ],
      "chunks": Array [
        0,
      ],
      "emitted": true,
      "info": Object {},
      "isOverSizeLimit": undefined,
      "name": "main.js",
      "size": 995,
    },
    Object {
      "chunkNames": Array [
        "main",
      ],
      "chunks": Array [
        0,
      ],
      "emitted": true,
      "info": Object {
        "development": true,
      },
      "isOverSizeLimit": undefined,
      "name": "main.js.map",
      "size": 97,
    },
    Object {
      "chunkNames": Array [],
      "chunks": Array [],
      "emitted": true,
      "info": Object {},
      "isOverSizeLimit": undefined,
      "name": "manifest.json",
      "size": 70,
    },
  ],
  "assetsByChunkName": Object {
    "main": Array [
      "main.js",
      "main.js.map",
    ],
  },
  "builtAt": 1482363367071,
  "children": Array [],
  "chunks": Array [
    Object {
      "children": Array [],
      "childrenByOrder": Object {},
      "entry": true,
      "files": Array [
        "main.js",
        "main.js.map",
      ],
      "filteredModules": 0,
      "hash": "3a071c49",
      "id": 0,
      "initial": true,
      "modules": Array [
        Object {
          "assets": Array [],
          "built": true,
          "cacheable": true,
          "chunks": Array [
            0,
          ],
          "depth": 0,
          "errors": 0,
          "failed": false,
          "id": 0,
          "identifier": "<PROJECT_ROOT>/node_modules/babel-loader/lib/index.js!<PROJECT_ROOT>/example/app.js",
          "index": 0,
          "index2": 0,
          "issuer": null,
          "issuerId": null,
          "issuerName": null,
          "issuerPath": null,
          "name": "./example/app.js",
          "optimizationBailout": Array [
            "ModuleConcatenation bailout: Module is not an ECMAScript module",
          ],
          "optional": false,
          "prefetched": false,
          "profile": undefined,
          "providedExports": null,
          "reasons": Array [
            Object {
              "explanation": undefined,
              "loc": "main",
              "module": null,
              "moduleId": null,
              "moduleIdentifier": null,
              "moduleName": null,
              "type": "single entry",
              "userRequest": "<PROJECT_ROOT>/example/app.js",
            },
          ],
          "size": 27,
          "source": "console.log('Hello world');",
          "usedExports": true,
          "warnings": 0,
        },
      ],
      "names": Array [
        "main",
      ],
      "origins": Array [
        Object {
          "loc": "main",
          "module": "",
          "moduleId": undefined,
          "moduleIdentifier": "",
          "moduleName": "",
          "reasons": Array [],
          "request": "<PROJECT_ROOT>/example/app.js",
        },
      ],
      "parents": Array [],
      "reason": undefined,
      "recorded": undefined,
      "rendered": true,
      "siblings": Array [],
      "size": 27,
    },
  ],
  "entrypoints": Object {
    "main": Object {
      "assets": Array [
        "main.js",
        "main.js.map",
      ],
      "childAssets": Object {},
      "children": Object {},
      "chunks": Array [
        0,
      ],
      "isOverSizeLimit": undefined,
    },
  },
  "errors": Array [],
  "filteredAssets": 0,
  "hash": Any<String>,
  "logging": Object {
    "webpack.buildChunkGraph.visitModules": Object {
      "debug": false,
      "entries": Array [],
      "filteredEntries": 2,
    },
  },
  "namedChunkGroups": Object {
    "main": Object {
      "assets": Array [
        "main.js",
        "main.js.map",
      ],
      "childAssets": Object {},
      "children": Object {},
      "chunks": Array [
        0,
      ],
      "isOverSizeLimit": undefined,
    },
  },
  "outputPath": "/dist",
  "publicPath": "/dist/",
  "time": 0,
  "version": "4.43.0",
  "warnings": Array [
    "configuration
The 'mode' option has not been set, webpack will fallback to 'production' for this value. Set 'mode' option to 'development' or 'production' to enable defaults for each environment.
You can also set it to 'none' to disable any default behavior. Learn more: https://webpack.js.org/configuration/mode/",
  ],
}
`;

exports[`fs 2`] = `
Array [
  "/dist/main.js",
  "/dist/main.js.map",
  "/dist/manifest.json",
]
`;
